trace/es/elasticsearch.go

```diff
@@ -0,0 +1,185 @@
+package es
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"runtime"
+	"strings"
+
+	"github.com/beatlabs/patron/trace"
+	"github.com/elastic/go-elasticsearch/v8"
+	"github.com/elastic/go-elasticsearch/v8/esapi"
+	"github.com/elastic/go-elasticsearch/v8/estransport"
+	"github.com/opentracing/opentracing-go"
+)
+
+const (
+	// ES Env Vars
+	DefaultHostEnv = "PATRON_ES_DEFAULT_HOST"
+	DefaultPortEnv = "PATRON_ES_DEFAULT_PORT"
+
+	// If no addresses are passed, hit localhost by default
+	DefaultHost = "http://localhost"
+
+	// If no ports are passed, hit default ES port by ...default
+	DefaultPort = "9200"
+
+	// span tag that denotes the host that responded to a query
+	respondentTag = "respondent"
+)
+
+type tracingInfo struct {
+	user  string
+	hosts []string
+}
+
+func (t *tracingInfo) startSpan(opName string, req *http.Request) (opentracing.Span, error) {
+
+	if req == nil {
+		return nil, fmt.Errorf("request is empty")
+	}
+
+	uri := req.URL.RequestURI()
+	method := req.Method
+
+	var bodyFmt string
+	if req.Body != nil {
+		if rawBody, err := ioutil.ReadAll(req.Body); err == nil {
+			req.Body = ioutil.NopCloser(bytes.NewReader(rawBody))
+			bodyFmt = string(rawBody)
+		}
+	}
+
+	return trace.EsSpan(req.Context(), opName, "go-elasticsearch", t.user, uri, method, bodyFmt, t.hosts), nil
+}
+
+func endSpan(sp opentracing.Span, rsp *http.Response) {
+	// Fetch the host that responded and inject it to the span
+	sp.SetTag(respondentTag, rsp.Request.URL.Host)
+
+	trace.FinishHTTPSpan(sp, rsp.StatusCode)
+}
+
+// transportClient is ES transport struct wrapped with tracing info
+type transportClient struct {
+	client *estransport.Client
+	tracingInfo
+}
+
+func (c *transportClient) Perform(req *http.Request) (*http.Response, error) {
+	// First try to retrieve the name of the caller to pass opName
+	var opName string
+	pc, _, _, ok := runtime.Caller(1)
```

Comment:
Yes, Im aware of that issue, a solution that I initially thought of, is adding a config that shuts down this functionality for cases where performance is crucial. Considering the fact that an ES client that is used for analytics will most times handle requests asynchronously.

I use reflection here because it doesn't run on a top down approach. I want to avoid wrapping the whole API under tracing functionality (this wouldn't scale also anyway) and have a ton of copy and paste code polluting the implementation. So on transport level, apart from access to the request I got no way to know which call is it.

The info that *is* known is the request path, body and method which in our case is maybe enough provided that the ES API  in practice isn't anything more than an HTTP API. So, if there isn't any better solution I think I will just remove this functionality and place a generic "ES call" instead.

Reply:
there are two options here I believe and correct me if I am wrong.
1. We query data 
2. We manipulate date (create, update, delete)
Maybe we should support these two operation names in some way.
If it is too much we can add the generic name and see later after usage if we need something more concrete.

Reply:
Some aggregations are POST while others and simple queries are GET so even this distinction isn't that straightforward. We can always parse request uri/body or have the caller inject metadata in the context but I'm not sure we really want either of these.